File: Server/blockBuddy.php
<?php
  require "globalFunctions.php";
  $pdo = new PDO("pgsql:host=".HOST.";port=".PORT.";dbname=".DATABASE.";user=".USERNAME.";password=".PASSWORD);

  $myId = $_POST["myId"];
  $token = $_POST["token"];
  $buddyId = $_POST["buddyId"];

	$isValid =
		isAuthenticated($myId, $token) &&
		isExtantUserId($buddyId);
	if (!$isValid) {
  	$pdo = null;
		die("Invalid");
	}

	$query = "select count(user_id) = 1 from account where user_id = ? and became_admin_at is not null limit 1;";
	$stmt = $pdo->prepare($query);
	$stmt->execute(array($buddyId));

  $row = $stmt->fetch();
	$isAdmin = $row[0];

	if ($isAdmin) {
    $return = array("isAdmin" => True);
  	echo json_encode($return);
		$pdo = null;
		exit;
	}

  $query = "select count(block_id) = 1 from block where user_id = ? and buddy_id = ? limit 1;";
  $stmt = $pdo->prepare($query);
  $stmt->execute(array($myId, $buddyId));

  $row = $stmt->fetch();
  $isBlocked = $row[0];

	// Buddy might have already been blocked from another device.
  if (!$isBlocked) {
  	$query = "insert into block (user_id, buddy_id) values (?, ?);";
  	$stmt = $pdo->prepare($query);
  	$stmt->execute(array($myId, $buddyId));
  }

	$pdo = null;
?>

File: Server/blockBuddy.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

func blockBuddy(w http.ResponseWriter, r *http.Request) {
	r.ParseForm()
	token := r.Form.Get("token")
	buddyId := r.Form.Get("buddyId")
	myId := getMyId(token)

	isValid :=
		isAuthenticated(myId) &&
			isExtantUserId(buddyId)
	if !isValid {
		fmt.Fprintln(w, "Invalid")
		return
	}

	if isAdmin(buddyId) {
		mapper := map[string]interface{}{
			"isAdmin": true,
		}
		json, err := json.Marshal(mapper)
		handleError(err)

		w.Header().Set("Content-Type", "application/json")
		w.Write(json)
		return
	}

	var isBlocked bool
	err := pool.QueryRow(context.Background(), "select count(block_id) = 1 from block where user_id = $1 and buddy_id = $2 limit 1;", myId, buddyId).Scan(&isBlocked)
	handleError(err)

	if !isBlocked {
		_, err := pool.Exec(context.Background(), "insert into block (user_id, buddy_id) values ($1, $2);", myId, buddyId)
		handleError(err)
	}
}
